File: Source/Runtime/CPlusplus/Foundation/Bibim/RectangleStorage.cpp
#include <Bibim/PCH.h>
#include <Bibim/RectangleStorage.h>
#include <Bibim/Assert.h>
#include <algorithm>

namespace Bibim
{
    RectangleStorage::RectangleStorage()
        : width(0),
          height(0)
    {
    }

    RectangleStorage::RectangleStorage(int width, int height)
        : width(width),
          height(height)
    {
        Clear(width, height);
    }

    RectangleStorage::RectangleStorage(const RectangleStorage& original)
        : width(original.width),
          height(original.height),
          freeRectangles(original.freeRectangles),
          allocatedRectangles(original.allocatedRectangles)
    {
    }

    RectangleStorage::~RectangleStorage()
    {
    }

    void RectangleStorage::Clear()
    {
        Clear(width, height);
    }

    void RectangleStorage::Clear(int width, int height)
    {
        BBAssert(width > 0 && height > 0);

        this->width  = width;
        this->height = height;
        freeRectangles.clear();
        freeRectangles.push_back(Rectangle(0, 0, width, height));
        allocatedRectangles.clear();
    }

    Rectangle RectangleStorage::Allocate(int width, int height)
    {
        if (GetWidth() == 0 || GetHeight() == 0)
            return Rectangle::Empty;

        // 1. 빈 영역중에 크기 같은 영역이 있으면 바로 할당한다.
        // 2. 빈 영역중에 가로크기나, 세로크기가 같은것이 있으면 분할해 할당한다.
        // 3. 빈 영역중 가장 가까운 영역을 찾아서 분할해 할당한다.

        Rectangle nearest = Rectangle::Empty;
        RectangleCollection::iterator nearestIterator = freeRectangles.end();

        for (RectangleCollection::iterator it = freeRectangles.begin(); it != freeRectangles.end(); ++it)
        {
            const Rectangle current = (*it);

            if (current.Width == width && current.Height == height)
            {
                freeRectangles.erase(it);
                allocatedRectangles.push_back(current);
                return current;
            }

            if (current.Width == width && current.Height >= height)
            {
                Rectangle result = current;
                result.Height = height;
                (*it).Y += height;
                (*it).Height -= height;
                allocatedRectangles.push_back(result);
                return result;
            }

            if (current.Width >= width && current.Height == height)
            {
                Rectangle result = current;
                result.Width = width;
                (*it).X += width;
                (*it).Width -= width;
                allocatedRectangles.push_back(result);
                return result;
            }

            if (current.Width >= width && current.Height >= height)
            {
                if (nearest == Rectangle::Empty || (current.Width <= nearest.Width && current.Height <= nearest.Height))
                {
                    nearest = current;
                    nearestIterator = it;
                }
            }
        }

        if (nearest != Rectangle::Empty)
        {
            const int xPoints[3] =
            {
                nearest.GetLeft(),
                nearest.GetLeft() + width,
                nearest.GetRight(),
            };
            const int yPoints[3] =
            {
                nearest.GetTop(),
                nearest.GetTop() + height,
                nearest.GetBottom(),
            };

            const Rectangle leftUp  = Rectangle(Point(xPoints[0], yPoints[0]), Point(xPoints[1], yPoints[1]));
            const Rectangle rightUp = Rectangle(Point(xPoints[1], yPoints[0]), Point(xPoints[2], yPoints[1]));
            const Rectangle down    = Rectangle(Point(xPoints[0], yPoints[1]), Point(xPoints[2], yPoints[2]));

            freeRectangles.erase(nearestIterator);
            freeRectangles.push_back(rightUp);
            freeRectangles.push_back(down);
            allocatedRectangles.push_back(leftUp);

            return leftUp;
        }
        else
            return Rectangle::Empty;
    }

    void RectangleStorage::Deallocate(const Rectangle& rectangle)
    {
        RectangleCollection::iterator it = std::find(allocatedRectangles.begin(), allocatedRectangles.end(), rectangle);
        if (it != allocatedRectangles.end())
        {
            freeRectangles.push_front(rectangle);
            Merge(freeRectangles, freeRectangles.begin());
            allocatedRectangles.erase(it);
        }
    }

    void RectangleStorage::Merge(RectangleCollection& rectangles, RectangleCollection::iterator targetIterator)
    {
        // targetIterator와 통합 가능한 영역을 찾아서 통합하고,
        // 그렇게 통합된 영역을 다른 영역과 통합해보기 위해 Merge를 재귀호출한다.

        const Rectangle target = (*targetIterator);
        for (RectangleCollection::iterator it = rectangles.begin(); it != rectangles.end(); ++it)
        {
            if (it == targetIterator)
                continue;

            Rectangle& current = *it;

            if (current.GetLeft() == target.GetLeft() && current.GetRight() == target.GetRight())
            {
                if (current.GetTop() == target.GetBottom())
                {
                    current.SetTop(target.GetTop());
                    rectangles.erase(targetIterator);
                    Merge(rectangles, it);
                    return;
                }

                if (current.GetBottom() == target.GetTop())
                {
                    current.SetBottom(target.GetBottom());
                    rectangles.erase(targetIterator);
                    Merge(rectangles, it);
                    return;
                }
            }

            if (current.GetTop() == target.GetTop() && current.GetBottom() == target.GetBottom())
            {
                if (current.GetLeft() == target.GetRight())
                {
                    current.SetLeft(target.GetLeft());
                    rectangles.erase(targetIterator);
                    Merge(rectangles, it);
                    return;
                }

                if (current.GetRight() == target.GetLeft())
                {
                    current.SetRight(target.GetRight());
                    rectangles.erase(targetIterator);
                    Merge(rectangles, it);
                    return;
                }
            }
        }
    }

    int RectangleStorage::GetWidth() const
    {
        return width;
    }

    int RectangleStorage::GetHeight() const
    {
        return height;
    }

    const RectangleStorage::RectangleCollection& RectangleStorage::GetFreeRectangles() const
    {
        return freeRectangles;
    }

    const RectangleStorage::RectangleCollection& RectangleStorage::GetAllocatedRectangles() const
    {
        return allocatedRectangles;
    }

    RectangleStorage& RectangleStorage::operator = (const RectangleStorage& right)
    {
        width = right.width;
        height = right.height;
        freeRectangles = right.freeRectangles;
        allocatedRectangles = right.allocatedRectangles;
        return *this;
    }

    bool RectangleStorage::operator == (const RectangleStorage& right) const
    {
        return freeRectangles == right.freeRectangles &&
               allocatedRectangles == right.allocatedRectangles;
    }

    bool RectangleStorage::operator != (const RectangleStorage& right) const
    {
        return !operator == (right);
    }
}

File: trunk/Source/Toolkit/Halak.Bibim.Toolkit/RectangleStorage.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Text;

namespace Halak.Bibim
{
    public sealed class RectangleStorage
    {
        #region Fields
        private LinkedList<Rectangle> allocatedRectangles;
        private LinkedList<Rectangle> freeRectangles;
        #endregion

        #region Properties
        public int Width
        {
            get;
            private set;
        }

        public int Height
        {
            get;
            private set;
        }

        public ICollection<Rectangle> AllocatedRectangles
        {
            get { return allocatedRectangles; }
        }

        public ICollection<Rectangle> FreeRectangles
        {
            get { return freeRectangles; }
        }
        #endregion

        #region Constructors
        public RectangleStorage(int width, int height)
        {
            Trace.Assert(width > 0 && height > 0);

            Width = width;
            Height = height;
            allocatedRectangles = new LinkedList<Rectangle>();
            freeRectangles = new LinkedList<Rectangle>();
        }

        public RectangleStorage(RectangleStorage original)
        {
            Trace.Assert(original != null);

            Width = original.Width;
            Height = original.Height;
            allocatedRectangles = new LinkedList<Rectangle>(original.allocatedRectangles);
            freeRectangles = new LinkedList<Rectangle>(original.freeRectangles);
        }
        #endregion

        #region Methods
        public void Clear()
        {
            allocatedRectangles.Clear();
            freeRectangles.Clear();
            freeRectangles.AddLast(new Rectangle(0, 0, Width, Height));
        }

        public Rectangle Allocate(int width, int height)
        {
            if (width == 0 || height == 0)
                return Rectangle.Empty;

            // 1. 빈 영역중에 크기 같은 영역이 있으면 바로 할당합니다.
            // 2. 빈 영역중에 가로크기나, 세로크기가 같은것이 있으면 분할해 할당합니다.
            // 3. 빈 영역중 가장 가까운 영역을 찾아서 분할해 할당합니다.

            Rectangle nearest = Rectangle.Empty;
            LinkedListNode<Rectangle> nearestNode = null;

            for (LinkedListNode<Rectangle> node = freeRectangles.First; node != null; node = node.Next)
            {
                Rectangle current = node.Value;
                if (current.Width == width && current.Height == height)
                {
                    freeRectangles.Remove(node);
                    allocatedRectangles.AddLast(current);
                    return current;
                }

                if (current.Width == width && current.Height >= height)
                {
                    Rectangle result = current;
                    result.Height = height;

                    node.Value = new Rectangle(current.X, current.Y + height,
                                               current.Width, current.Height - height);
                    allocatedRectangles.AddLast(result);

                    return result;
                }

                if (current.Width >= width && current.Height == height)
                {
                    Rectangle result = current;
                    result.Width = width;

                    node.Value = new Rectangle(current.X + width, current.Y,
                                               current.Width - width, current.Height);
                    allocatedRectangles.AddLast(result);

                    return result;
                }

                if (current.Width >= width && current.Height >= height)
                {
                    if (nearest.IsEmpty || (current.Width <= nearest.Width && current.Height <= nearest.Height))
                    {
                        nearest = current;
                        nearestNode = node;
                    }
                }
            }

            if (nearest.IsEmpty == false)
            {
                int xPoint1 = nearest.Left;
                int xPoint2 = nearest.Left + width;
                int xPoint3 = nearest.Right;
                int yPoint1 = nearest.Top;
                int yPoint2 = nearest.Top + height;
                int yPoint3 = nearest.Bottom;

                Rectangle leftUp = Rectangle.FromLTRB(xPoint1, yPoint1, xPoint2, yPoint2);
                Rectangle rightUp = Rectangle.FromLTRB(xPoint2, yPoint1, xPoint3, yPoint2);
                Rectangle down = Rectangle.FromLTRB(xPoint1, yPoint2, xPoint3, yPoint3);

                freeRectangles.Remove(nearestNode);
                freeRectangles.AddLast(rightUp);
                freeRectangles.AddLast(down);
                allocatedRectangles.AddLast(leftUp);

                return leftUp;
            }
            else
                return Rectangle.Empty;
        }

        public void Deallocate(Rectangle rectangle)
        {
            LinkedListNode<Rectangle> node = allocatedRectangles.Find(rectangle);
            if (node != null)
            {
                freeRectangles.AddLast(rectangle);
                Merge(freeRectangles, freeRectangles.First);
                allocatedRectangles.Remove(node);
            }
            else
                Trace.WriteLine("RectangleStorage.Deallocate > 할당되지 않은 Rect를 Deallocate 하였습니다. ({0})", rectangle.ToString());
        }

        public static void Merge(LinkedList<Rectangle> rectangles, LinkedListNode<Rectangle> targetNode)
        {
            // targetIterator와 통합 가능한 영역을 찾아서 통합하고,
            // 그렇게 통합된 영역을 다른 영역과 통합해보기 위해 Merge를 재귀호출합니다.

            Rectangle target = targetNode.Value;
            for (LinkedListNode<Rectangle> node = rectangles.First; node != null; node = node.Next)
            {
                if (node == targetNode)
                    continue;

                Rectangle current = node.Value;

                if (current.Left == target.Left && current.Right == target.Right)
                {
                    if (current.Top == target.Bottom)
                    {
                        node.Value = Rectangle.FromLTRB(current.Left, target.Top, current.Right, current.Bottom);
                        rectangles.Remove(targetNode);
                        Merge(rectangles, node);
                        return;
                    }

                    if (current.Bottom == target.Top)
                    {
                        node.Value = Rectangle.FromLTRB(current.Left, current.Top, current.Right, target.Bottom);
                        rectangles.Remove(targetNode);
                        Merge(rectangles, node); 
                        return;
                    }
                }

                if (current.Top == target.Bottom && current.Bottom == target.Bottom)
                {
                    if (current.Left == target.Right)
                    {
                        node.Value = Rectangle.FromLTRB(target.Left, current.Top, current.Right, current.Bottom);
                        rectangles.Remove(targetNode);
                        Merge(rectangles, node);
                        return;
                    }

                    if (current.Right == target.Left)
                    {
                        node.Value = Rectangle.FromLTRB(current.Left, current.Top, target.Right, target.Bottom);
                        rectangles.Remove(targetNode);
                        Merge(rectangles, node);
                        return;
                    }
                }
            }
        }
        #endregion
    }
}
